clippy_lints/src/loops.rs

```diff
@@ -424,47 +422,100 @@ impl<'a, 'tcx> LateLintPass<'a, 'tcx> for Pass {
+fn never_loop(block: &Block, id: &NodeId) -> bool {
+    !contains_continue_block(block, id) && loop_exit_block(block)
+}
+
+fn contains_continue_block(block: &Block, dest: &NodeId) -> bool {
+    block.stmts.iter().any(|e| contains_continue_stmt(e, dest))
+        || block.expr.as_ref().map_or(false, |e| contains_continue_expr(e, dest))
 }
 
-fn never_loop_stmt(stmt: &Stmt) -> bool {
+fn contains_continue_stmt(stmt: &Stmt, dest: &NodeId) -> bool {
     match stmt.node {
         StmtSemi(ref e, _) |
-        StmtExpr(ref e, _) => never_loop_expr(e),
-        StmtDecl(ref d, _) => never_loop_decl(d),
+        StmtExpr(ref e, _) => contains_continue_expr(e, dest),
+        StmtDecl(ref d, _) => contains_continue_decl(d, dest),
     }
 }
 
-fn never_loop_decl(decl: &Decl) -> bool {
-    if let DeclLocal(ref local) = decl.node {
-        local.init.as_ref().map_or(false, |e| never_loop_expr(e))
-    } else {
-        false
+fn contains_continue_decl(decl: &Decl, dest: &NodeId) -> bool {
+    match decl.node {
+        DeclLocal(ref local) => local.init.as_ref().map_or(false, |e| contains_continue_expr(e, dest)),
+        _ => false
     }
 }
 
-fn never_loop_expr(expr: &Expr) -> bool {
+fn contains_continue_expr(expr: &Expr, dest: &NodeId) -> bool {
     match expr.node {
-        ExprBreak(..) | ExprRet(..) => true,
         ExprBox(ref e) |
         ExprUnary(_, ref e) |
-        ExprBinary(_, ref e, _) | // because short-circuiting
         ExprCast(ref e, _) |
         ExprType(ref e, _) |
         ExprField(ref e, _) |
         ExprTupField(ref e, _) |
-        ExprRepeat(ref e, _) |
-        ExprAddrOf(_, ref e) => never_loop_expr(e),
+        ExprAddrOf(_, ref e) |
+        ExprRepeat(ref e, _) => contains_continue_expr(e, dest),
+        ExprBinary(_, ref e1, ref e2) |
         ExprAssign(ref e1, ref e2) |
         ExprAssignOp(_, ref e1, ref e2) |
-        ExprIndex(ref e1, ref e2) => never_loop_expr(e1) || never_loop_expr(e2),
+        ExprIndex(ref e1, ref e2) => [e1, e2].iter().any(|e| contains_continue_expr(e, dest)),
         ExprArray(ref es) |
         ExprTup(ref es) |
-        ExprMethodCall(_, _, ref es) => es.iter().any(|e| never_loop_expr(e)),
-        ExprCall(ref e, ref es) => never_loop_expr(e) || es.iter().any(|e| never_loop_expr(e)),
-        ExprBlock(ref block) => never_loop_block(block),
-        ExprStruct(_, _, ref base) => base.as_ref().map_or(false, |e| never_loop_expr(e)),
+        ExprMethodCall(_, _, ref es) => es.iter().any(|e| contains_continue_expr(e, dest)),
+        ExprCall(ref e, ref es) => contains_continue_expr(e, dest) || es.iter().any(|e| contains_continue_expr(e, dest)),
+        ExprBlock(ref block) => contains_continue_block(block, dest),
+        ExprStruct(_, _, ref base) => base.as_ref().map_or(false, |e| contains_continue_expr(e, dest)),
+        ExprAgain(d) => d.target_id.opt_id().map_or(false, |id| id == *dest),
```

Comment:
I'm not sure. `opt_id()` only returns `None` in cases where there is a [LoopIdError](https://manishearth.github.io/rust-internals-docs/rustc/hir/enum.LoopIdError.html) which means the `continue` is invalid. So I figured if the `continue` is invalid then I should ignore it.